django/db/models/sql/compiler.py

```diff
@@ -903,17 +905,102 @@ def __init__(self, *args, **kwargs):
+        # of '%s' and [None]. The 'NULL' placeholder is produced earlier by
+        # OracleOperations.get_geom_placeholder(). The following line removes
+        # the corresponding None parameter. See ticket #10888.
+        params = self.connection.ops.modify_insert_params(sql, params)
+
+        return sql, params
+
+    def prepare_value(self, field, value):
+        """
+        Prepare a value to be used in a query, by resolving it if it is an
+        expression, and otherwise calling the field's get_db_prep_save().
+        """
+
+        def find_fs(expr):
+            """Walk this expression tree depth-first and yield all F nodes."""
+            if isinstance(expr, F):
+                yield expr
+            elif isinstance(expr, CombinedExpression):
+                for e in chain(find_fs(expr.lhs), find_fs(expr.rhs)):
+                    yield e
+
+        if hasattr(value, 'resolve_expression'):
+            # Don't allow F() expressions. They refer to existing columns on a
+            # row, but in the case of insert the row doesn't exist yet.
+            if any(find_fs(value)):
+                raise ValueError(
+                    'Failed to insert expression "%s" on %s. F() expressions '
+                    'can only be used to update, not to insert.' % (value, field)
+                )
+            value = value.resolve_expression(self.query, allow_joins=False, for_save=True)
+            if value.contains_aggregate:
+                raise FieldError("Aggregate functions are not allowed in this query")
+        else:
+            value = field.get_db_prep_save(value, connection=self.connection)
+        return value
+
+    def pre_save_val(self, field, obj):
+        """
+        Get the given field's value off the given obj. pre_save() is used for
+        things like auto_now on DateTimeField. Skip it if this is a raw query.
+        """
+        if self.query.raw:
+            return getattr(obj, field.attname)
+        return field.pre_save(obj, add=True)
+
+    def assemble_as_sql(self, fields, value_rows):
+
+        if not value_rows:
+            return [], []
+
+        # list of (sql, [params]) tuples for each object to be saved
+        # Shape: [n_objs][n_fields][2]
+        rows_of_fields_as_sql = (
+            (self.field_as_sql(field, v) for field, v in zip(fields, row))
+            for row in value_rows
+        )
+
+        # tuple like ([sqls], [[params]s]) for each object to be saved
+        # Shape: [n_objs][2][n_fields]
+        sql_and_param_pair_rows = (zip(*row) for row in rows_of_fields_as_sql)
+
+        # Extract separate lists for placeholders and params.
+        # Each of these has shape [n_objs][n_fields]
+        placeholder_rows, param_rows = zip(*sql_and_param_pair_rows)
+
+        # Params for each field are still lists, and need to be flattened.
+        flatten = lambda lists: [item for lst in lists for item in lst]
```

Comment:
This could be extracted too. Also it's a pep8 warning on the latest pep8 - apparently you shouldn't assign lambdas and should use `def` instead.